server/src/main/java/org/elasticsearch/cluster/coordination/ValidateJoinRequest.java

```diff
@@ -7,32 +7,96 @@
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.compress.CompressorFactory;
+import org.elasticsearch.common.io.stream.InputStreamStreamInput;
+import org.elasticsearch.common.io.stream.NamedWriteableAwareStreamInput;
+import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.core.RefCounted;
+import org.elasticsearch.core.internal.io.IOUtils;
 import org.elasticsearch.transport.TransportRequest;
 
 import java.io.IOException;
 
 public class ValidateJoinRequest extends TransportRequest {
-    private ClusterState state;
+    private final CheckedSupplier<ClusterState, IOException> stateSupplier;
+    private final RefCounted refCounted;
 
     public ValidateJoinRequest(StreamInput in) throws IOException {
         super(in);
-        this.state = ClusterState.readFrom(in, null);
+        if (in.getVersion().onOrAfter(Version.V_8_3_0)) {
+            // recent versions send a BytesTransportRequest containing the compressed state
+            final var bytes = in.readReleasableBytesReference();
+            final var version = in.getVersion();
+            final var namedWriteableRegistry = in.namedWriteableRegistry();
+            this.stateSupplier = () -> readCompressed(version, bytes, namedWriteableRegistry);
+            this.refCounted = bytes;
+        } else {
+            // older versions just contain the bare state
+            final var state = ClusterState.readFrom(in, null);
+            this.stateSupplier = () -> state;
+            this.refCounted = null;
+        }
+    }
+
+    private static ClusterState readCompressed(Version version, BytesReference bytes, NamedWriteableRegistry namedWriteableRegistry)
+        throws IOException {
+        final var compressor = CompressorFactory.compressor(bytes);
+        StreamInput in = bytes.streamInput();
+        try {
+            if (compressor != null) {
+                in = new InputStreamStreamInput(compressor.threadLocalInputStream(in));
+            }
+            in = new NamedWriteableAwareStreamInput(in, namedWriteableRegistry);
+            in.setVersion(version);
+            try (StreamInput input = in) {
+                return ClusterState.readFrom(input, null);
+            }
+        } finally {
+            IOUtils.close(in);
+        }
     }
 
     public ValidateJoinRequest(ClusterState state) {
-        this.state = state;
+        this.stateSupplier = () -> state;
+        this.refCounted = null;
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
+        assert out.getVersion().before(Version.V_8_3_0);
         super.writeTo(out);
-        this.state.writeTo(out);
+        stateSupplier.get().writeTo(out);
+    }
+
+    public ClusterState getState() throws IOException {
+        return stateSupplier.get();
```

Comment:
It is only called once so I think this is fine. I renamed it in fd908df6b82b7f12105b044f5f2b2528cc8dd66a to `getOrReadState` just to make it a bit clearer that this might be more than a simple getter.